docs/tutorials-redeliver-messages.md

```diff
@@ -0,0 +1,236 @@
+
+There might be instances where the received messages cannot be processed at this time or some errors happened during processing.
+The client needs to redeliver the unacknowledged messages or a particular message after a delay or immediately.
+Pulsar provides at-least-once semantics when the client does not enable transaction because the client may cache some
+messages out of Pulsar when redelivering messages.
+
+Pulsar Consumer API provides four ways to reconsume the unacknowledged messages later:
+
+* **ackTimeout**
+* **deadLetterPolicy**
+  * **reconsumeLaterCumulative**
+  * **reconsumeLater**
+* **negativeAcknowledge**
+* **redeliverUnacknowledgedMessages**
+
+The **ackTimeout** is an automatic protection mechanism. If a consumer configured ackTimeout, the messages will be
+auto-redelivered when the received messages are not acknowledged after a long time. It works at the client side, ensuring
+the unacknowledged messages will be redelivered to another consumer when the connection of a consumer remains active
+but the business system gets stuck or misses to ack some messages. If the consumer gets disconnected when the client crashes,
+the messages will be auto-redelivered by the broker too. However, this mechanism does not wait for the response of acknowledgment,
+so if an acknowledgment fails on the broker side or proxy side, an ack hole may occur.
+
+The **deadLetterPolicy** is a policy in the message queue used to handle messages that cannot be processed properly.
+In many message queue systems (such as RabbitMQ, Google Pub/Sub, Apache Pulsar, etc.), this strategy is implemented.
+In Pulsar, the deadLetterPolicy is implemented at the client side, it creates a new retry letter topic and dead letter
+topic when building a consumer with `deadLetterPolicy` configuration. When a consumer calls `reconsumeLaterCumulative`
+or `reconsumeLater`, the message (method parameter) will be produced to the retry letter topic until the retry time reaches
+the `maxRedeliverCount`. The message will be produced to the dead letter topic when the retry time reaches the `
+maxRedeliverCount`. The main difference between them is that `reconsumeLaterCumulative` will cumulative ack the
+message (method parameter) after it is produced and `reconsumeLater` will individual ack the message (method parameter)
+after it is produced.
+
+The `negativeAcknowledge` is used to redeliver certain unacknowledged messages while `redeliverUnacknowledgedMessages`
+is used to redeliver all the unacknowledged messages received by this consumer. The main difference between them and
+deadLetterPolicy is that there is no new topic created, and there is an unlimited number of redeliveries.
+
+## Best Practice Suggestion
+
+Different scenarios require different best practices. Users who value the order of partition messages and wish to batch
+process data should choose or implement an appropriate routerPolicy to send a batch of ordered messages to the same
```

Comment:
You're talking about producer best practices here? If so, I think it's out of scope for Redelivery content, no?

Reply:
This is related to the choice of subscription type. If users wish to consume a batch of messages in order, and they choose the exclusive mode or failover mode, but they do not send this batch of messages to the same partition, then these messages will be consumed out of order. This is something users do not want to see. To avoid confusing users, it is necessary to explain that before choosing exclusive and failover modes for ordered message consumption, they should ensure that those messages that need to be consumed in order have been sent to the same partition. 

We have not elaborated on how to implement the router. This is still a lead-up to the best practices for consumption.